modules/nextflow/src/main/groovy/nextflow/script/parser/v2/ScriptCompiler.groovy

```diff
@@ -0,0 +1,295 @@
+            ? unit.getSource(MAIN_CLASS_NAME, codeSource.getFile())
+            : unit.getSource(codeSource.getName(), codeSource.getScriptText())
+        final collector = new ScriptClassLoader(loader).createCollector(unit, su)
+
+        compiler = new Compiler(unit)
+        compiler.addSource(su)
+
+        unit.addSource(su)
+        unit.setClassgenCallback(collector)
+        unit.compile(Phases.CLASS_GENERATION)
+
+        // collect script classes
+        final classes = collector.getLoadedClasses().stream()
+            .map((o) -> (Class) o)
+            .filter((c) -> c.getSuperclass() == BaseScript.class)
+            .toList()
+
+        // extract main script class
+        final main = classes.stream()
+            .filter(c -> c.getSimpleName() == MAIN_CLASS_NAME)
+            .findFirst()
+            .get()
+
+        // match each module script class to the source path
+        // using the class name
+        final modules = new HashMap<Path,Class>()
+        for( final c : classes ) {
+            for( final source : unit.getModules() ) {
+                if( source.getName() == c.getSimpleName() ) {
+                    final path = Path.of(source.getSource().getURI())
+                    modules.put(path, c)
+                    break
+                }
+            }
+        }
+        return new CompileResult(main, modules)
+    }
+
+    public static record CompileResult(
+        Class main,
+        Map<Path,Class> modules
+    ) {}
+
+    private static class ScriptClassLoader extends GroovyClassLoader {
+        public ScriptClassLoader(GroovyClassLoader parent) {
+            super(parent)
+        }
+
+        @Override
+        public ClassCollector createCollector(CompilationUnit unit, SourceUnit su) {
+            return super.createCollector(unit, su)
+        }
+    }
+
+    private class ScriptCompilationUnit extends CompilationUnit {
+
+        private static final List<ClassNode> DEFAULT_IMPORTS = List.of(
+            new ClassNode( java.nio.file.Path ),
+            new ClassNode( nextflow.Channel ),
+            new ClassNode( nextflow.util.Duration ),
+            new ClassNode( nextflow.util.MemoryUnit ),
+        )
+
+        private SourceUnit entry
+
+        private Set<SourceUnit> modules
+
+        public ScriptCompilationUnit(CompilerConfiguration configuration, GroovyClassLoader loader) {
+            super(configuration, null, loader)
+            super.addPhaseOperation(source -> analyze(source), Phases.CONVERSION)
+        }
+
+        public Set<SourceUnit> getModules() {
+            return modules
+        }
+
+        @Override
+        public void addPhaseOperation(final ISourceUnitOperation op, final int phase) {
+            super.addPhaseOperation((source) -> {
+                // skip main script on second conversion pass
+                if( phase == Phases.CONVERSION && source == entry )
+                    return
+                op.call(source)
+            }, phase)
```

Comment:
I think the IDE is correct to report an ambiguity. Even if it compiles it's not clear what method is invoking


![Uploading Screenshot 2025-03-10 at 18.20.41.png…]()

Reply:
I figured out how to keep the script compiler in Java, so these IntelliJ errors should go away.

I had to re-do some commits to fix the DCO, so you might have to reset and re-pull
```bash
git reset --hard 6cd60d1
git pull
```